staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -472,16 +472,21 @@ func (w *watchCache) WaitUntilFreshAndList(ctx context.Context, resourceVersion
 		return nil, 0, "", err
 	}
 
-	// This isn't the place where we do "final filtering" - only some "prefiltering" is happening here. So the only
-	// requirement here is to NOT miss anything that should be returned. We can return as many non-matching items as we
-	// want - they will be filtered out later. The fact that we return less things is only further performance improvement.
-	// TODO: if multiple indexes match, return the one with the fewest items, so as to do as much filtering as possible.
-	for _, matchValue := range matchValues {
-		if result, err := w.store.ByIndex(matchValue.IndexName, matchValue.Value); err == nil {
-			return result, w.resourceVersion, matchValue.IndexName, nil
+	result, rv, index, err := func() ([]interface{}, uint64, string, error) {
+		// This isn't the place where we do "final filtering" - only some "prefiltering" is happening here. So the only
+		// requirement here is to NOT miss anything that should be returned. We can return as many non-matching items as we
+		// want - they will be filtered out later. The fact that we return less things is only further performance improvement.
+		// TODO: if multiple indexes match, return the one with the fewest items, so as to do as much filtering as possible.
+		for _, matchValue := range matchValues {
+			if result, err := w.store.ByIndex(matchValue.IndexName, matchValue.Value); err == nil {
+				return result, w.resourceVersion, matchValue.IndexName, nil
+			}
 		}
-	}
-	return w.store.List(), w.resourceVersion, "", nil
+		return w.store.List(), w.resourceVersion, "", nil
+	}()
+
+	// FIXME: The result is not sorted when returned by watchcache (as opposed to etcd).
```

Comment:
This is the first bigger issue that we found due to this effort of generalizing tests across all implementation of storage interface (there are three other small fixed in this PR).

@liggitt - FYI

Reply:
Why does it matter if the result is sorted or not from a `LIST` call?

Reply:
This is good question.
From one hand, this is what users get when the list from etcd (etcd returns a sorted list of results). Given that this was original semantic, people may potentially be used for it.
OTOH - it clearly wasn't the case in watchcache, since it was created. Given that noone realized this discrepancy, it may also mean that it's not a critical feature. Lack of sorting is also beneficial from performance perspective, so if this isn't needed, I would love to not have it. But consistency seems more important to me.

I just added the sorting to this PR to make the semantic consistency with etcd implemention - if we decide (before merging this PR) that this isn't needed, removing it (I consciously did it in a separate commit) and adjusting tests is trivial.

I would love to hear from @liggitt about this and probably also from @lavalamp and @deads2k

Reply:
https://github.com/kubernetes/kubernetes/pull/108392 helps with the sorting needs (should it be required, in fact the BTree approach is more or less the way etcd does it internally as well)(definitely an overkill if _only_ sort-ability is required).

Reply:
We end up requiring a sorted collection to implement paginated responses, right?

Reply:
> I would love to hear from @liggitt about this and probably also from @lavalamp and @deads2k

This is in my queue, but I probably can't get to it this week. Is this trying to make it into 1.26?

Reply:
I think this was a deliberate choice when we made the watch cache. However I support fixing it to match etcd, especially since we need to do that anyway to paginate.

Reply:
I'm surprised the two do not match.  I would make the watch cache consistent with etcd.

Reply:
> I'm surprised the two do not match. I would make the watch cache consistent with etcd.

That's what this PR is doing in this matter.